pymatgen/analysis/diffraction/tem.py

```diff
@@ -0,0 +1,788 @@
+import plotly.graph_objs as go
+import plotly.offline as poff
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.widgets import Cursor, Button
+from IPython.display import set_matplotlib_formats, Image, display
+set_matplotlib_formats('retina')
+poff.init_notebook_mode(connected=True)
+
+# coding: utf-8
+# Copyright (c) Pymatgen Development Team.
+# Distributed under the terms of the MIT License.
+from math import *
+from fractions import Fraction 
+
+import os
+import json
+import math
+import numpy as np
+import random
+
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
+
+from pymatgen.analysis.diffraction.core import DiffractionPattern, AbstractDiffractionPatternCalculator, \
+    get_unique_families
+
+from pymatgen.analysis.local_env import site_is_of_motif_type, MinimumDistanceNN
+
+with open(os.path.join(os.path.dirname(__file__),
+                       "atomic_scattering_params.json")) as f:
+    ATOMIC_SCATTERING_PARAMS = json.load(f)
+
+"""
+This module implements a TEM pattern calculator.
+"""
+
+#Credit to Dr. Shyue Ping Ong for the template of the calculator
+__author__ = "Frank Wan"
+__copyright__ = "Copyright 2018, The Materials Project"
+__version__ = "0.1"
+__maintainer__ = "Frank Wan respect for S.P.O"
+__email__ = "fwan@berkeley.edu"
+__date__ = "06/19/2019"
+
+class TEMDot():
+    """
+    Instantatiates a point on the TEM diffraction pattern.
+    """
+# TODO: make this MSONable
+# TODO: add type hints for every method
+# TODO: memoize methods to improve runtime
+# TODO: standardize input argument convention (i.e., should things like bragg_angles be passed in, or calculated ad hoc?)
+# TODO: add support for showing different Laue zones.
+    def __init__(self, position: List[float], hkl: List[int], intensity: float, film_radius: float, d_spacing: float) -> None:
+        """
+        Args:
+              hkl (3-tuple): The hkl plane that the point corresponds/is indexed to.
+              d_spacing (float): The interplanar spacing of the dot.
+              film_radius (float): The radius of the dot on the film. Determined 
+                  by microscope aberration equations (ie Cs corrections and other such
+                  aberrations)
+              intensity (float): The intensity of the dot. Determines its brightness
+                  in the pattern, relative to those of the other dots.
+              position (Position): The xy-coordinates of the dot on the plot.
+        """
+        self.position = position
+        self.hkl = hkl
+        self.intensity = intensity    
+        self.film_radius = film_radius        
+        self.d_spacing = d_spacing
+
+class TEMCalculator(AbstractDiffractionPatternCalculator):
+    """
+    Computes the TEM pattern of a crystal structure for multiple Laue zones.
+    """
+
+    """
+    some important constants needed for calculations
+    """
+    planck = 6.62607004 * 10**-34
+    restMassElec = 9.1093835611 * 10**-31
```

Comment:
Please follow pymatgen's convention for variable names, i.e., underscores, not CamelCase.